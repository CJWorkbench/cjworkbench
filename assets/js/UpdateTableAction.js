import {store, addModuleAction, setParamValueAction, setParamValueActionByIdName, setSelectedWfModuleAction} from './workflow-reducer'
import {findParamValByIdName} from './utils'

// Unit tests written in individual test.js files per module ex: SortFromTable.test.js

// Map Module id_name to update function and moduleId in workflows.py
export const updateModuleMapping = {
  'selectcolumns':    updateSelectModule,
  'duplicate-column': updateDuplicateModule,
  'filter':           updateFilterModule,
  'editcells':        addEditToEditCellsModule,
  'rename-columns':   updateRenameModule,
  'reorder-columns':  updateReorderModule,
  'sort-from-table':  updateSortModule,
  'extract-numbers':  updateExtractNumbersModule,
  'clean-text':       updateCleanTextModule
}

// Constants for sort module
const SortTypes = 'text|number|datetime'.split('|')
export const sortDirectionNone = 0
export const sortDirectionAsc = 1
export const sortDirectionDesc = 2

// Constants for select module
export const selectColumnDrop = 0
export const selectColumnKeep = 1

// Constant for extract numbers module
export const extractNumberAny = 0

function findModuleIdByIdName (state, moduleIdName) {
  return state.updateTableModuleIds[moduleIdName] || null
}

// Find if a module of moduleId exists as or is next to module with focusWfModuleId
function findModuleWithIds (state, focusWfModuleId, moduleId) {
  const { workflow, wfModules } = state

  // validIdsOrNulls: [ 2, null, null, 65 ] means indices 0 and 3 are for
  // desired module (and have wfModuleIds 2 and 64), 1 and 2 aren't for
  // desired module
  const validIdsOrNulls = workflow.wf_modules
    .map(id => (wfModules[String(id)].module_version || {}).module === moduleId ? id : null)

  const focusIndex = workflow.wf_modules.indexOf(focusWfModuleId)
  if (focusIndex === -1) return null

  // Are we already focused on a valid WfModule?
  const atFocusIndex = validIdsOrNulls[focusIndex]
  if (atFocusIndex !== null) return wfModules[String(atFocusIndex)]

  // Is the _next_ wfModule valid? If so, return that
  const nextIndex = focusIndex + 1
  const atNextIndex = validIdsOrNulls[nextIndex]
  if (atNextIndex !== null) return wfModules[String(atNextIndex)]

  // Nope, no target module with moduleIdName where we need it
  return null
}

function ensureSelectedWfModule (state, wfModule) {
  const current = state.selected_wf_module
  let wanted = state.workflow ? state.workflow.wf_modules.indexOf(wfModule.id) : null
  if (wanted === -1) wanted = null

  if (wanted !== null && wanted !== current) {
    store.dispatch(setSelectedWfModuleAction(wanted));
  }
}

export function updateTableActionModule (wfModuleId, idName, forceNewModule, params) {
  const state = store.getState()
  // Check if module imported
  const moduleId = findModuleIdByIdName(state, idName)
  if (moduleId === null) {
    window.alert("Module '" + idName + "' not imported.")
    return
  }
  const existingModule = findModuleWithIds(state, wfModuleId, moduleId)
  if (existingModule && !forceNewModule) {
    ensureSelectedWfModule(state, existingModule) // before state's existingModule changes
    updateModuleMapping[idName](existingModule, params) // ... changing state's existingModule
  } else {
    const wfModuleIndex = state.workflow.wf_modules.indexOf(wfModuleId)
    store.dispatch(addModuleAction(moduleId, wfModuleIndex + 1))
      .then(fulfilled => {
        const newWfm = fulfilled.value.data.wfModule
        updateModuleMapping[idName](newWfm, params)
      })
  }
}

// TODO this can be the model for a default multicolumn handler.
// Will need to pass module name
function updateCleanTextModule (wfm, params) {
  let existingColumns = findParamValByIdName(wfm, 'colnames')
  let newColumn = params.columnKey

  if (existingColumns.value) {
    // Do nothing if column already exists
    if (existingColumns.value.split(',').includes(newColumn)) {}
    else {updateTableActionModule(wfm.id, 'clean-text', true, params)}
  }
  else {
    store.dispatch(setParamValueActionByIdName(wfm.id, 'colnames', params.columnKey))
  }
}

function updateExtractNumbersModule (wfm, params) {
  let extractedColumns = findParamValByIdName(wfm, 'colnames')

  if (extractedColumns.value) {
    let extract = findParamValByIdName(wfm, 'extract').value
    let formatType = findParamValByIdName(wfm, 'type_format').value
    let replaceType = findParamValByIdName(wfm, 'type_replace').value
    // If column already an existing module parameter, do nothing
    if (extractedColumns.value.split(',').includes(params.columnKey)) {}
    // If existing module not defaults, force new
    else if ((extract) | (formatType !== 0) | (replaceType !== 0)) {
      updateTableActionModule(wfm.id, 'extract-numbers', true, params)
    }
    else {
      let newColumns = extractedColumns.value + ',' + params.columnKey
      store.dispatch(setParamValueActionByIdName(wfm.id, 'colnames', newColumns))
    }
  }
  // New module with extract type 'Any'
  else {
    store.dispatch(setParamValueActionByIdName(wfm.id, 'colnames', params.columnKey))
    store.dispatch(setParamValueActionByIdName(wfm.id, 'type', extractNumberAny))
  }
}

function updateFilterModule (wfm, params) {
  store.dispatch(setParamValueActionByIdName(wfm.id, 'column', params.columnKey))
}

function updateSelectModule (wfm, params) {
  let selectedColumns = findParamValByIdName(wfm, 'colnames')
  const action = findParamValByIdName(wfm, 'drop_or_keep')
  const dropColumnName = params.columnKey

  // Case: If module exists and drop already selected
  if (selectedColumns.value && action.value === selectColumnDrop) {
    if (!(selectedColumns.value.split(',').includes(dropColumnName))) {
      let entries = selectedColumns.value + ',' + dropColumnName
      store.dispatch(setParamValueActionByIdName(wfm.id, 'colnames', entries))
    }
  }
  // Case: If module exists and keep already selected, deselect dropColumnName
  else if (selectedColumns.value && action.value === selectColumnKeep) {
    selectedColumns = selectedColumns.value.split(',')
    let dropColumnNameIdx = selectedColumns.indexOf(dropColumnName)
    if (dropColumnNameIdx > -1) {
      selectedColumns.splice(dropColumnNameIdx, 1)
      store.dispatch(setParamValueActionByIdName(wfm.id, 'colnames', selectedColumns.toString()))
      store.dispatch(setParamValueActionByIdName(wfm.id, 'drop_or_keep', selectColumnKeep))
    }
  }
  // Case: If no existing module
  else if (!selectedColumns.value) {
    store.dispatch(setParamValueActionByIdName(wfm.id, 'colnames', dropColumnName))
    store.dispatch(setParamValueActionByIdName(wfm.id, 'drop_or_keep', selectColumnDrop))
  }
}

function addEditToEditCellsModule (wfm, params) {
  const param = findParamValByIdName(wfm, 'celledits')
  const edit = params
  let edits

  if (param.value) {
    try {
      edits = JSON.parse(param.value)
      // Remove the previous edit to the same cell
      let idx = edits.findIndex((element) => {
        return element.row === params.row && element.col === params.col
      })
      if (idx > -1) {
        edits.splice(idx, idx + 1)
      }
    } catch (err) {
      console.error(err)
      edits = []
    }
  } else {
    edits = []
  }

  // Add this edit and update the server
  edits.push(edit)
  store.dispatch(setParamValueAction(param.id, JSON.stringify(edits)))
}

function updateDuplicateModule (wfm, params) {
  const entriesParam = findParamValByIdName(wfm, 'colnames')
  const duplicateColumnName = params.columnKey

  // if params already exist, check if duplicateColumnName already exists
  if (entriesParam.value) {
    if (!(entriesParam.value.split(',').includes(duplicateColumnName))) {
      let entries = entriesParam.value + ',' + duplicateColumnName
      store.dispatch(setParamValueActionByIdName(wfm.id, 'colnames', entries))
    }
    // if duplicateColumnName already in entriesParam, do nothing
  } else {
    store.dispatch(setParamValueActionByIdName(wfm.id, 'colnames', duplicateColumnName))
  }
}

function updateSortModule (wfm, params) {
  // Must be kept in sync with sortfromtable.json
  const sortTypeIdx = SortTypes.indexOf(params.sortType)
  store.dispatch(setParamValueActionByIdName(wfm.id, 'column', params.columnKey))
  store.dispatch(setParamValueActionByIdName(wfm.id, 'direction', params.sortDirection))
  store.dispatch(setParamValueActionByIdName(wfm.id, 'dtype', sortTypeIdx))
}

// renameInfo format: {prevName: <current column name in table>, newName: <new name>}

function updateRenameModule (wfm, params) {
  const entriesParam = findParamValByIdName(wfm, 'rename-entries')
  let entries

  if (entriesParam.value) {
    try {
      entries = JSON.parse(entriesParam.value)
    } catch (e) {
      console.warn(e)
      entries = {}
    }
  } else {
    entries = {}
  }
  // If "prevName" in renameInfo exists as a value in edit entries,
  // update that entry (since we are renaming a renamed column)
  let entryExists = false
  for (let k in entries) {
    if (entries[k] === params.prevName) {
      entries[k] = params.newName
      entryExists = true
      break
    }
  }
  // Otherwise, add the new entry to existing entries.
  if (!entryExists) {
    entries[params.prevName] = params.newName
  }
  store.dispatch(setParamValueAction(entriesParam.id, JSON.stringify(entries)))
}

function updateReorderModule (wfm, params) {
  var historyParam = findParamValByIdName(wfm, 'reorder-history')
  var historyStr = historyParam ? historyParam.value.trim() : ''
  var historyEntries = []

  try {
    historyEntries = JSON.parse(historyStr)
  } catch (e) {
    // Something is wrong with our history. Erase it and start over, what else can we do?
  }

  // User must drag two spaces to indicate moving one column right (because drop = place before this column)
  // So to prevent all other code from having to deal with this forever, decrement the index here
  if (params.to > params.from) {
    params.to -= 1
  }

  historyEntries.push(params)
  store.dispatch(setParamValueAction(historyParam.id, JSON.stringify(historyEntries)))
}
