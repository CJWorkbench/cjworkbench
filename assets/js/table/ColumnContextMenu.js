// Drop-down menu on Workflows List page, for each listed WF
// triggered by click on three-dot icon next to listed workflow

import React from 'react'
import PropTypes from 'prop-types'
import { UncontrolledDropdown, DropdownMenu, DropdownToggle, DropdownItem, DropdownDivider } from '../components/Dropdown'
import { Trans } from '@lingui/macro'

export default class ColumnContextMenu extends React.Component {
  static propTypes = {
    onClickAction: PropTypes.func.isRequired, // func(idName, forceNewModule, params)
    columnType: PropTypes.string.isRequired,
    renameColumn: PropTypes.func.isRequired
  }

  createOrUpdate (idName, extraParams = {}) {
    this.props.onClickAction(idName, false, extraParams)
  }

  create (idName, extraParams = {}) {
    this.props.onClickAction(idName, true, extraParams)
  }

  handleRenameColumn = (...args) => { this.props.renameColumn(...args) }
  handleDuplicateColumn = () => this.createOrUpdate('duplicatecolumns')
  handleSortAscending = () => this.createOrUpdate('sort', { is_ascending: true })
  handleSortDescending = () => this.createOrUpdate('sort', { is_ascending: false })
  handleAddNewFilter = () => this.create('filter')
  handleExtractNumbers = () => this.createOrUpdate('converttexttonumber')
  handleCleanText = () => this.createOrUpdate('clean-text')
  handleDropColumn = () => this.createOrUpdate('selectcolumns', { keep: false })
  handleConvertDate = () => this.createOrUpdate('convert-date')
  handleConvertText = () => this.createOrUpdate('converttotext')
  handleFormatNumbers = () => this.create('formatnumbers', { format: '{:,}' })

  render () {
    const { columnType } = this.props

    return (
      <UncontrolledDropdown>
        <DropdownToggle className='context-button'>
          <i className='icon-caret-down' />
        </DropdownToggle>
        <DropdownMenu>
          <DropdownItem onClick={this.handleRenameColumn} className='rename-column-header' icon='icon-edit'><Trans id='js.table.ColumnContextMenu.rename.'>Rename</Trans></DropdownItem>
          <DropdownItem onClick={this.handleDuplicateColumn} className='duplicatecolumns' icon='icon-duplicate'><Trans id='js.table.ColumnContextMenu.duplicate'>Duplicate</Trans></DropdownItem>
          <DropdownDivider />
          <DropdownItem onClick={this.handleSortAscending} className='sort-ascending' icon='icon-sort-up'><Trans id='js.table.ColumnContextMenu.sortAscending'>Sort ascending</Trans></DropdownItem>
          <DropdownItem onClick={this.handleSortDescending} className='sort-descending' icon='icon-sort-down'><Trans id='js.table.ColumnContextMenu.sortDescending'>Sort descending</Trans></DropdownItem>
          <DropdownDivider />
          <DropdownItem onClick={this.handleAddNewFilter} className='filter-column' icon='icon-filter'><Trans id='js.table.ColumnContextMenu.filter'>Filter</Trans></DropdownItem>
          <DropdownItem onClick={this.handleCleanText} className='clean-text' icon='icon-clean'><Trans id='js.table.ColumnContextMenu.cleanText'>Clean Text</Trans></DropdownItem>
          <DropdownDivider />
          <DropdownItem onClick={this.handleConvertDate} className='convert-date' icon='icon-calendar'><Trans id='js.table.ColumnContextMenu.convertToDateTime'>Convert to date & time</Trans></DropdownItem>
          <DropdownItem onClick={this.handleExtractNumbers} className='converttexttonumber' icon='icon-number'><Trans id='js.table.ColumnContextMenu.convertToNumbers'>Convert to numbers</Trans></DropdownItem>
          <DropdownItem onClick={this.handleConvertText} className='converttotext' icon='icon-text'><Trans id='js.table.ColumnContextMenu.convertToText'>Convert to text</Trans></DropdownItem>
          <DropdownDivider />
          <DropdownItem onClick={this.handleFormatNumbers} className='formatnumbers' icon='icon-number' disabled={columnType !== 'number'}><Trans id='js.table.ColumnContextMenu.formatNumbers'>Format numbers</Trans></DropdownItem>
          <DropdownDivider />
          <DropdownItem onClick={this.handleDropColumn} className='drop-column' icon='icon-removec'><Trans id='js.table.ColumnContextMenu.deleteColumn'>Delete column</Trans></DropdownItem>
        </DropdownMenu>
      </UncontrolledDropdown>
    )
  }
}
