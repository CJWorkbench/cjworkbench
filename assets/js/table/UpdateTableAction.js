import { addModuleAction, setWfModuleParamsAction, setSelectedWfModuleAction } from '../workflow-reducer'

/**
 * Module param-building functions per id_name.
 *
 * Each function has the signature (oldParams, params), and it should
 * return new params or null.
 *
 * `oldParams` will be `null` if this is a new module.
 *
 * Watch out: these param builders are NOT all used in the same way! For
 * example, editcells' "params" is used to add an _edit_, not overwrite an
 * "edits" param. In effect, this is a bunch of functions with completely
 * different purposes but with the same calling convention (kinda like Redux
 * reducers.) Most are used in the column drop-down menu. TODO move the others
 * elsewhere.
 */
export const moduleParamsBuilders = {
  'selectcolumns': buildSelectColumnsParams,
  'duplicatecolumns': genericAddColumn('colnames', false),
  'filter': buildFilterParams,
  'editcells': buildEditCellsParams,
  'renamecolumns': buildRenameColumnsParams,
  'reordercolumns': buildReorderColumnsParams,
  'sort': buildSortColumnsParams,
  'converttexttonumber': genericAddColumn('colnames', false),
  'clean-text': genericAddColumn('colnames', true),
  'convert-date': genericAddColumn('colnames', false),
  'converttotext': genericAddColumn('colnames', true),
  'formatnumbers': genericAddColumn('colnames', true)
}

/**
 * Return { id, params, isNext } of _this_ WfModule or the one _after_ it, matching moduleIdName.
 *
 * Return `null` if this or the next WfModule does not match moduleIdName.
 */
function findWfModuleWithIds (state, focusWfModuleId, moduleIdName) {
  const { tabs, wfModules, modules } = state

  if (!(moduleIdName in modules)) {
    throw new Error(`Cannot find module '${moduleIdName}'`)
  }

  const wfModule = wfModules[String(focusWfModuleId)]
  const tabSlug = wfModule.tab_slug
  const tab = tabs[tabSlug]

  // validIdsOrNulls: [ 2, null, null, 65 ] means indices 0 and 3 are for
  // desired module (and have wfModuleIds 2 and 64), 1 and 2 aren't for
  // desired module
  const validIdsOrNulls = tab.wf_module_ids
    .map(id => wfModules[String(id)].module === moduleIdName ? id : null)

  const focusIndex = tab.wf_module_ids.indexOf(focusWfModuleId)
  if (focusIndex === -1) return null

  // Are we already focused on a valid WfModule?
  const atFocusIndex = validIdsOrNulls[focusIndex]
  if (atFocusIndex !== null) {
    const wfModule = wfModules[String(atFocusIndex)]
    return {
      id: atFocusIndex,
      params: wfModule.params,
      isNext: false
    }
  }

  // Is the _next_ wfModule valid? If so, return that
  const nextIndex = focusIndex + 1
  const atNextIndex = nextIndex >= validIdsOrNulls.length ? null : validIdsOrNulls[nextIndex]
  if (atNextIndex !== null) {
    const wfModule = wfModules[String(atNextIndex)]
    return {
      id: atNextIndex,
      params: wfModule.params,
      isNext: true
    }
  }

  // Nope, no target module with moduleIdName where we need it
  return null
}

/**
 * Adds or edits a module, given `wfModuleId` as the selected table.
 *
 * This is a reducer action that delegates to `addModuleAction`, `setSelectedWfModuleAction` and `
 */
export function updateTableAction (wfModuleId, idName, forceNewModule, params) {
  return (dispatch, getState) => {
    const state = getState()

    if (!(idName in state.modules)) {
      window.alert("Module '" + idName + "' not imported.")
      return
    }

    const existingWfModule = forceNewModule ? null : findWfModuleWithIds(state, wfModuleId, idName)
    const newParams = moduleParamsBuilders[idName](existingWfModule ? existingWfModule.params : null, params)

    if (existingWfModule && !forceNewModule) {
      if (existingWfModule.id !== wfModuleId) {
        dispatch(setSelectedWfModuleAction(existingWfModule.id))
      }
      if (newParams) {
        dispatch(setWfModuleParamsAction(existingWfModule.id, newParams))
      }
    } else {
      dispatch(addModuleAction(idName, { afterWfModuleId: wfModuleId }, newParams))
    }
  }
}

function buildSelectColumnsParams (oldParams, params) {
  if (Object.keys(params).length === 0) {
    // A bit of a hack: if we call this with no params, we're just adding
    // an empty select-columns module with default params.
    return {}
  }

  const colnames = oldParams ? oldParams.colnames.split(',').filter(s => !!s) : []
  const keep = params ? params.drop_or_keep == 1 : true
  const oldKeep = oldParams ? oldParams.drop_or_keep == 1 : keep
  const keepIdx = keep ? 1 : 0
  const oldKeepIdx = oldKeep ? 1 : 0
  const colname = params.columnKey // may be undefined

  if (colnames.length === 0) {
    // Adding a new module, or resetting an empty one
    return { colnames: colname, drop_or_keep: keepIdx }
  } else {
    const idx = colnames.indexOf(colname)

    if (oldKeep) {
      // Existing module is "keep"
      if (keep) {
        throw new Error('Unhandled case: trying to keep in a keep module')
        return null
      } else {
        // We want to drop
        if (idx === -1) {
          return null // the keep module doesn't include colname: it's already dropped
        } else {
          colnames.splice(idx, 1)
          return { colnames: colnames.join(','), drop_or_keep: oldKeepIdx }
        }
      }
    } else {
      // Existing module is "drop"
      if (keep) {
        throw new Error('Unhandled case: trying to keep in a remove module')
        return null
      } else {
        // We want to drop another one
        if (idx === -1) {
          colnames.push(colname)
          return { colnames: colnames.join(','), drop_or_keep: oldKeepIdx }
        } else {
          return null // the drop module is already dropping colname
        }
      }
    }

    if (keep === oldKeep) {
      if (keep && idx === -1) {
        colnames.push(params.columnKey)
        return { colnames, drop_or_keep: keepIdx }
      } else if (!keep && idx !== -1) {
      } else {
        return null
      }
    } else {
      console.warn('Unimplemented drop/keep logic', oldKeep, keep, colnames, colname)
      return null
    }
  }
}

function buildEditCellsParams (oldParams, params) {
  const edits = (oldParams && oldParams.celledits) ? oldParams.celledits : []
  const edit = params

  // Remove the previous edit to the same cell
  const idx = edits.findIndex(({ row, col }) => row === edit.row && col === edit.col)

  if (idx === -1) {
    edits.push(edit)
    return { celledits: edits }
  } else if (edits[idx].value !== edit.value) {
    edits.splice(idx, 1, edit)
    return { celledits: edits }
  } else {
    return null
  }
}

function buildFilterParams (_oldParams, params) {
  return {
    filters: {
      operator: 'and',
      filters: [
        {
          operator: 'and',
          subfilters: [
            {
              colname: params.columnKey,
              condition: '',
              value: '',
              case_sensitive: false
            }
          ]
        }
      ]
    }
  }
}

function genericSetColumn (key) {
  return (oldParams, params) => {
    const newParams = { ...params }
    const colname = newParams.columnKey
    if (newParams.hasOwnProperty('columnKey')) {
      delete newParams.columnKey
      newParams[key] = colname
    }
    return newParamsUnlessNoChange(oldParams, newParams)
  }
}

function newParamsUnlessNoChange (oldParams, newParams) {
  if (!oldParams) return newParams
  for (const key in oldParams) {
    if (oldParams[key] !== newParams[key]) {
      return newParams
    }
  }
  return null
}

function genericAddColumn (key, deprecatedStringStorage) {
  return (oldParams, params) => {
    let colnames
    if (deprecatedStringStorage) {
      colnames = oldParams ? (oldParams[key] || '').split(',').filter(s => !!s) : []
    } else {
      colnames = oldParams ? (oldParams[key] || []) : []
    }
    if (!params.hasOwnProperty('columnKey'))  throw new Error('Expected "columnKey" column to add')
    const colname = params.columnKey

    if (!colname) throw new Error('Unexpected params: ' + JSON.stringify(params))

    if (colnames.includes(colname)) {
      return null
    } else {
      const newParams = { ...params }
      const colname = newParams.columnKey
      delete newParams.columnKey
      const newColnames = [ ...colnames, colname ]
      newParams[key] = deprecatedStringStorage ? newColnames.join(',') : newColnames

      return newParamsUnlessNoChange(oldParams, newParams)
    }
  }
}

function buildRenameColumnsParams (oldParams, params) {
  // renameInfo format: {prevName: <current column name in table>, newName: <new name>}
  const renames = oldParams && oldParams.renames || {}
  const { prevName, newName } = params

  if (renames[prevName] === newName) {
    return null
  } else {
    return {
      renames: {
        ...renames,
        [prevName]: newName
      }
    }
  }
}

function buildReorderColumnsParams (oldParams, params) {
  // Yes, yes, this is half-broken -- https://www.pivotaltracker.com/story/show/162592381
  const historyEntries = (oldParams && oldParams['reorder-history']) ? JSON.parse(oldParams['reorder-history']) : []

  let { column, to, from } = params

  // User must drag two spaces to indicate moving one column right (because drop = place before this column)
  // So to prevent all other code from having to deal with this forever, decrement the index here
  if (to > from) {
    to -= 1
  }

  historyEntries.push({ column, to, from })
  return { 'reorder-history': JSON.stringify(historyEntries) }
}

function buildSortColumnsParams (oldParams, params) {
  // 1. If the column is already the first param and directions the same, return null
  // 2. Remove existing column from param list, if it exists
  // 3. Prepend list with new column
  const newColumn = { colname: params.columnKey, is_ascending: params.is_ascending }

  if (oldParams == null) {
    return { sort_columns: [newColumn], keep_top: '' }
  }

  const columns = oldParams.sort_columns

  if (columns[0].colname === newColumn.colname && columns[0].is_ascending === newColumn.is_ascending) {
    return null
  }

  const sort_columns = columns.filter( column => column.colname !== params.columnKey)
  sort_columns.unshift(newColumn)

  const newParams = { sort_columns: sort_columns, keep_top: oldParams.keep_top }

  return newParams
}
