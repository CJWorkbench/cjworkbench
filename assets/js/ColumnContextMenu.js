// Drop-down menu on Workflows List page, for each listed WF
// triggered by click on three-dot icon next to listed workflow

import React from 'react'
import PropTypes from 'prop-types'
import Portal from 'reactstrap/lib/Portal'
import UncontrolledDropdown from 'reactstrap/lib/UncontrolledDropdown'
import DropdownToggle from 'reactstrap/lib/DropdownToggle'
import DropdownMenu from 'reactstrap/lib/DropdownMenu'
import DropdownItem from 'reactstrap/lib/DropdownItem'
import { sortDirectionNone, sortDirectionAsc, sortDirectionDesc} from './UpdateTableAction'

// Modifiers disabled to prevent menu flipping (occurs even when flip=false)
var dropdownModifiers = {
  preventOverflow: {
    enabled: false
  },
  hide: {
    enabled: false
  }
}

export default class ColumnContextMenu extends React.Component {
  static propTypes = {
    setDropdownAction: PropTypes.func.isRequired,
    columnKey:         PropTypes.string.isRequired,
    columnType:        PropTypes.string.isRequired,
    renameColumn:      PropTypes.func.isRequired,
    sortDirection:     PropTypes.oneOf([sortDirectionNone, sortDirectionAsc, sortDirectionDesc]).isRequired
  }

  // Modules that only need to pass the column name and do not force new modules use setDropdownActionDefault
  setDropdownActionDefault = (idName) => this.props.setDropdownAction(idName, false, {})
  setDropdownActionForceNew = (idName) => this.props.setDropdownAction(idName, true, {})
  setSortDirection (direction) {
    let params = {
      'sortType': this.props.columnType,
      'sortDirection': direction
    }
    this.props.setDropdownAction('sort-from-table', false, params)
  }

  render() {
    return (
      <UncontrolledDropdown>
        <DropdownToggle className='context-button'>
          <i className='icon-more'></i>
        </DropdownToggle>
        <Portal>
          <DropdownMenu persist flip={false} modifiers={dropdownModifiers}>
            <DropdownItem onClick={this.props.renameColumn} className='rename-column-header' toggle={false}>
              <i className="icon-edit"></i>
              <span>Rename</span>
            </DropdownItem>
            <DropdownItem onClick={this.setDropdownActionDefault.bind(this, 'duplicate-column')} className='duplicate-column' toggle={false}>
              <i className="icon-duplicate"></i>
              <span>Duplicate</span>
            </DropdownItem>
            <DropdownItem divider />
            <DropdownItem onClick={this.setSortDirection.bind(this, sortDirectionAsc)} className='sort-ascending' toggle={false}>
              <i className="icon-sort-up"></i>
              <span>Sort ascending</span>
            </DropdownItem>
            <DropdownItem onClick={this.setSortDirection.bind(this, sortDirectionDesc)} className='sort-descending' toggle={false}>
              <i className="icon-sort-down"></i>
              <span>Sort descending</span>
            </DropdownItem>
            <DropdownItem divider />
            <DropdownItem onClick={this.setDropdownActionForceNew.bind(this, 'filter')} className='filter-column' toggle={false}>
              <i className="icon-filter"></i>
              <span>Filter</span>
            </DropdownItem>
            <DropdownItem onClick={this.setDropdownActionDefault.bind(this, 'selectcolumns')} className='drop-column' toggle={false}>
              <i className="icon-removec"></i>
              <span>Drop column</span>
            </DropdownItem>
            <DropdownItem onClick={this.setDropdownActionDefault.bind(this, 'extract-numbers')} className='extract-numbers' toggle={false}>
              <i className="icon-number"></i>
              <span>Extract Numbers</span>
            </DropdownItem>
          </DropdownMenu>
        </Portal>
      </UncontrolledDropdown>
    );
  }
}
