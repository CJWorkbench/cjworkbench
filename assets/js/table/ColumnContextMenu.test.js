/* global describe, it, jest, expect */
import React from 'react'
import ColumnContextMenu from './ColumnContextMenu'
import { shallow } from 'enzyme'
import { sortDirectionNone, sortDirectionAsc, sortDirectionDesc } from './UpdateTableAction'

describe('ColumnContextMenu', () => {
  function mountMenu (setDropdownAction, columnKey, columnType, sortDirection, renameColumn) {
    return shallow(
      <ColumnContextMenu
        setDropdownAction={setDropdownAction}
        columnKey={columnKey}
        columnType={columnType}
        sortDirection={sortDirection}
        renameColumn={renameColumn}
      />
    )
  }

  it('should match snapshot', () => {
    let wrapper = mountMenu(jest.fn(), 'columnKey', 'text', sortDirectionNone, jest.fn())
    expect(wrapper).toMatchSnapshot() // stores file which represents tree of component
  })

  // only checking the call to drop down functions, not actual actions on table
  it('should call functions ', async () => {
    const setDropdownAction = jest.fn()
    const renameColumn = jest.fn()

    const wrapper = mountMenu(setDropdownAction, 'columnKey', 'text', sortDirectionNone, renameColumn)

    wrapper.find('DropdownItem.drop-column').simulate('click')
    expect(setDropdownAction).toHaveBeenCalledWith('selectcolumns', false, {})

    wrapper.find('DropdownItem.rename-column-header').simulate('click')
    expect(renameColumn).toHaveBeenCalled()

    wrapper.find('DropdownItem.duplicate-column').simulate('click')
    expect(setDropdownAction).toHaveBeenCalledWith('duplicate-column', false, {})

    wrapper.find('DropdownItem.filter-column').simulate('click')
    expect(setDropdownAction).toHaveBeenCalledWith('filter', true, {})

    wrapper.find('DropdownItem.sort-ascending').simulate('click')
    expect(setDropdownAction).toHaveBeenCalledWith('sort-from-table', false, {sortType: 'text', sortDirection: sortDirectionAsc})

    wrapper.find('DropdownItem.sort-descending').simulate('click')
    expect(setDropdownAction).toHaveBeenCalledWith('sort-from-table', false, {sortType: 'text', sortDirection: sortDirectionDesc})
  })
})
