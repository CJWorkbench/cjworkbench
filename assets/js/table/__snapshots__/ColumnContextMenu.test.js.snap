// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnContextMenu should match snapshot 1`] = `
<ColumnContextMenu
  columnKey="columnKey"
  columnType="text"
  onClickAction={[MockFunction]}
  renameColumn={[MockFunction]}
>
  <UncontrolledDropdown>
    <Dropdown
      isOpen={false}
      toggle={[Function]}
    >
      <Manager
        tag={false}
      >
        <div
          className="dropdown"
        >
          <DropdownToggle
            className="context-button"
          >
            <Target>
              <button
                aria-expanded={false}
                className="btn btn-secondary context-button"
                disabled={false}
                onClick={[Function]}
              >
                <i
                  className="icon-caret-down"
                />
              </button>
            </Target>
          </DropdownToggle>
          <DropdownMenu />
        </div>
      </Manager>
    </Dropdown>
  </UncontrolledDropdown>
</ColumnContextMenu>
`;
